Add new Navigator components and use them in the global styles sidebar

youknowriad opened the pull request:
borrows from #32923 and refactors the navigation in global styles sidebar from #34885 to use the new Navigator components. This is more flexible in terms of design though it requires a bit more code.

youknowriad committed:
Use the new navigator components

youknowriad committed:
Simplify the Navigation components implementation

jorgefilipecosta commented on packages/components/src/navigator/match-path.js:
```diff
@@ -0,0 +1,79 @@
+}
+
+/**
+ * Public API for matching a URL pathname to a path.
```
The method is not public/exported outside the package right? Should we replace "Public API" with function?

youknowriad replied:
I'm actually considering removing this function/file for now. All what we really need for now is `path === currentPath` we can consider advanced use-cases later.

jorgefilipecosta commented on packages/components/src/navigator/screen.js:
```diff
@@ -0,0 +1,81 @@
+const animationExitDuration = 0.14;
+const animationExitDelay = 0;
+
+function Screen( {
```
Personally, I would prefer if the component was named NavigatorScreen instead of Screen. We use `<NavigatorScreen />` everywhere so naming it NavigatorScreen makes it easier to associate mentally that it is this component being referenced.

jorgefilipecosta commented on packages/components/src/navigator/screen.js:
```diff
@@ -0,0 +1,81 @@
+
+function Screen( {
+	children,
+	exact = false,
```
Is there any case where we expect non-exact matches to happen?
Is the plan to allow something like rendering this NavigatorScreen on all paths "/blocks/..." in the future?

youknowriad replied:
We could but yeah for now I think we can remove all the complex path matching https://github.com/WordPress/gutenberg/pull/34904#discussion_r712769222

jorgefilipecosta commented on packages/components/src/navigator/match-path.js:
```diff
@@ -0,0 +1,79 @@
+ * @param {string|Array} options  Matching options.
+ * @return {Object} Matched path details
+ */
+function matchPath( pathname, options = {} ) {
```
I'm missing some context on this function.
I guess when isExact is true we verify if pathname is equal to options.path if yes we return true. This is the most common case, should we have a special condition to avoid regex, etc on this case and just compare the strings? (it seems a good performance improvement).

When isExact is false what patches do we match? Is it just a path like "blocks/" matches all paths with shape "blocks/.../...". If yes do we need a regex for that? Could we just verify if the string starts with "blocks/"? I guess the issue is much more complex and we accept a diverse set of matches but I'm missing context on what we are supposed to accept and not so I'm having trouble properly reviewing the algorithm.

jorgefilipecosta replied:
matchPath is the type of function that may benefit from some units tests. The unit tests help document the function and better understand what it does, and may also allow future changes easier.

jorgefilipecosta commented on packages/components/src/navigator/match-path.js:
```diff
@@ -0,0 +1,79 @@
+ * Public API for matching a URL pathname to a path.
+ *
+ * @param {string}       pathname Pathname
+ * @param {string|Array} options  Matching options.
```
I guess here we should document we accept {string|Array|Object}. And the documentation should say something like the path to be matched against pathName or an object specifying the path to be matched and the matching options.

jorgefilipecosta commented on packages/edit-site/src/components/sidebar/global-styles-sidebar.js:
```diff
@@ -8,14 +8,28 @@ import { map } from 'lodash';
  */
 import {
 	Button,
-	__experimentalNavigation as Navigation,
```
What's the plan regarding these components? Are they going to be refactored and use __experimentalNavigator inside?

youknowriad replied:
That's my hope yes. (I'd appreciate some help there with existing usage :P )

jorgefilipecosta commented on packages/edit-site/src/components/sidebar/global-styles-sidebar.js:
```diff
@@ -186,63 +250,73 @@ export default function GlobalStylesSidebar() {
-				</NavigationMenu>
+				</NavigatorScreen>
+
 				{ map( blocks, ( block, name ) => (
```
For blocks that don't support any global styles panel, we should not render a menu for them. I think we regressed on https://github.com/WordPress/gutenberg/pull/34885.

youknowriad replied:
We can leave this separate I think to avoid having a diff unrelated with the change of the navigation components.

jorgefilipecosta reviewed (approved):
Nice work on this PR @youknowriad. Things are working well for me 👍  I left some questions related mostly to our matching mechanism. In general, I like this approach it is simple and I think it seems like a good fit for our needs. I guess we can merge this PR.

youknowriad committed:
Simpler path matching

ciampo commented:
> I guess we can merge this PR.

I would prefer to wait before merging until we have a clear understanding of the approach we're going to take with the navigation components if that's possible

youknowriad committed:
Added a README and a story

jorgefilipecosta commented on packages/components/package.json:
```diff
@@ -58,6 +58,7 @@
 		"lodash": "^4.17.21",
 		"memize": "^1.1.0",
 		"moment": "^2.22.1",
+		"path-to-regexp": "^0.1.7",
```
We are not using this module anymore, right? Should we remove it?

youknowriad committed:
Some lingering diffs

youknowriad committed:
Tweak styles

ciampo reviewed (approved):
Thank you for iterating on this new set of components, @youknowriad ! Overall the code looks much cleaner, and it decouples the UI from the functionality.

I just wanted to mention that in both the GLobal Styles sidebar and the Storybook example, I can spot a horizontal scrollbar while two screens are animating at the same time. Here's a video (I slowed down the animation times to show this better):


https://user-images.githubusercontent.com/1083581/134344879-11eb5091-9790-4c9b-a6a6-b8e9bfd51290.mp4



Apart from the scrollbar, in my mind the next steps would be:
- convert the `Navigator` components to TypeScript
- add some basic a11y features:
  - focus management: when clicking on a link, the new `Screen` should receive focus
  - RTL support: on RTL languages, we should swap the direction of the animation
  - reduced motion: we should disable the animations completely in case the user has expressed a preference for reduced motion

youknowriad committed:
Support reduced motion

youknowriad committed:
Add RTL support

youknowriad committed:
Add focus on mount behavior

youknowriad commented:
Thanks for the review Marco.

> RTL support: on RTL languages, we should swap the direction of the animation
> reduced motion: we should disable the animations completely in case the user has expressed a preference for reduced motion

I implemented both of these.

>  I can spot a horizontal scrollbar while two screens are animating at the same time.

Good catch, it's not really clear how to solve and keep the same animation, adding a "overflow: hidden" could be a solution about I'm worried about side effects (hidden popovers potentially)

> convert the Navigator components to TypeScript

😬 I'm not yet familiar with that yet but I can try

> focus management: when clicking on a link, the new Screen should receive focus

I implemented a naive version of this. This might require some iterations or input from others.

youknowriad commented:
Actually the typescript conversion here has proven a bit challenging for me, I'd love to see someone else do it (and learn from it) either here or separately.

ciampo commented:
Thank you for working on the RTL, reduced motion and focus management — they all seem to work well for me and I think they make for a good initial set of a11y improvements. We should probably add a changelog entry to the `@wordpress/components` package mentioning the introduction of a new experimental set of `Navigator` components.

I suggest that we merge this PR, and work on a few follow-ups:

- conversion to **TypeScript** (I am happy to work on it this or next week, and I can tag both @youknowriad and @jorgefilipecosta in the PR so that you can follow along)
- fix the horizontal **scrollbar** showing while transitioning
- do a more in-depth **a11y audit** and work on any necessary improvements


What do folks think?

ciampo commented on packages/components/src/navigator/screen.js:
```diff
@@ -0,0 +1,92 @@
+	const ref = useFocusOnMount();
+
+	// This flag is used to only apply the focus on mount when the actual path changes.
+	// It avois the focus to happen on first render.
```
```suggestion
	// It avoids the focus to happen on the first render.
```

youknowriad committed:
Fix typo

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

youknowriad commented:
@ciampo This plan makes sense to me.

youknowriad committed:
Update the changelog

ciampo commented:
> @ciampo This plan makes sense to me.

Cool, I've updated the tracking issue #34907 with the next steps regarding the `Navigator` component
